### PR TITLE
Circular view dependency check

### LINKED ISSUE #1158: [FEATURE]: Check circular dependencies when migrating views
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem statement

Although theoretically made impossible by the database server, there exists a scenario where views could refer to each other:
```
CREATE VIEW V1 AS SELECT * FROM V2
CREATE VIEW V2 AS SELECT * FROM V1
```
This could happen if the DDL is extracted at different points in time and consolidated outside the database.
This would lead to an infinite loop in the ViewMigrator, and an impossibility to migrate altogether.
It is therefore useful to check for this scenario.

### Proposed Solution

Enhance `src/databricks/labs/ucx/hive_metastore/views_migrator.py` such that it checks for circular dependencies and eventually raises an error 

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by jincejames**
## Changes
Added check for circular view dependency

### Linked issues
fixes #1158 

Resolves #..

### Functionality 

- [ ] added relevant user documentation
- [ ] added new CLI command
- [ ] modified existing command: `databricks labs ucx ...`
- [ ] added a new workflow
- [ ] modified existing workflow: `...`
- [ ] added a new table
- [ ] modified existing table: `...`

### Tests
Exisiting test for circular view dependency is satisfied and a new test to check chained circular dependency (A->B, B->C, C->A) has been added.

- [ ] manually tested
- [x] added unit tests
- [ ] added integration tests
- [ ] verified on staging environment (screenshot attached)

### CODE REVIEW

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/hive_metastore/view_migrate.py [hunk 1]
```diff
@@ -127,5 +128,31 @@ def _next_batch(self, views: set[ViewToMigrate]) -> set[ViewToMigrate]:
...
+        queue.extend(dep for dep in initial_view.dependencies)
+        print(queue)
+        while queue:
+            current_view = self._get_view_instance(queue.pop(0).key, views)
+            if current_view:
+                if current_view == initial_view:
+                    raise ValueError(
+                        f"Circular dependency detected between {initial_view.src.name} and {current_view.src.name} "
+                    )
+                queue.extend(dep for dep in current_view.dependencies)
```
[the remark this answers is not in the document]
Comment (jincejames) [reply src/databricks/labs/ucx/hive_metastore/view_migrate.py]: removed nesting

File: src/databricks/labs/ucx/hive_metastore/view_migrate.py [hunk 2]
```diff
@@ -127,5 +128,31 @@ def _next_batch(self, views: set[ViewToMigrate]) -> set[ViewToMigrate]:
...
+        # any of the views in the list of views. This means the circular dependency will be identified only
+        # if the dependencies are present in the list of views passed to _next_batch() or the _result_view_list
+        # ToDo: see if a mapper between TableView and ViewToMigrate can be implemented
+        instance = None
+        all_views = list(views) + self._result_view_list
+        for view in all_views:
+            if view.src.key == key:
+                instance = view
+                break
+        return instance
```
[the remark this answers is not in the document]
Comment (jincejames) [reply src/databricks/labs/ucx/hive_metastore/view_migrate.py]: removed variable

File: src/databricks/labs/ucx/hive_metastore/view_migrate.py [hunk 3]
```diff
@@ -127,5 +128,31 @@ def _next_batch(self, views: set[ViewToMigrate]) -> set[ViewToMigrate]:
...
         # prevent infinite loop
         if len(result) == 0 and len(views) > 0:
-            raise ValueError(f"Circular view references are preventing migration: {views}")
+            raise ValueError(f"Invalid table references are preventing migration: {views}")
         return result
+
+    def check_circular_dependency(self, initial_view, views):
+        queue = []
+        queue.extend(dep for dep in initial_view.dependencies)
+        print(queue)
```
[the remark this answers is not in the document]
Comment (jincejames) [reply src/databricks/labs/ucx/hive_metastore/view_migrate.py]: Debug code removed

**Review by nfx — APPROVED** [review]
lgtm

**Pull request closed without merging**